src/Adapter/Product/CommandHandler/UpdateProductStatusHandler.php

```diff
@@ -66,19 +66,30 @@ public function __construct(
      */
     public function handle(UpdateProductStatusCommand $command)
     {
-        $product = $this->productRepository->get($command->getProductId());
+        $shopConstraint = $command->getShopConstraint();
+        $product = $this->productRepository->getByShopConstraint($command->getProductId(), $shopConstraint);
         $initialState = (bool) $product->active;
         $product->active = $command->getEnable();
         $this->productRepository->partialUpdate(
             $product,
             ['active'],
+            $shopConstraint,
             CannotUpdateProductException::FAILED_UPDATE_STATUS
         );
 
-        // If status changed we need to update its indexes (we check if it is necessary because index build can be
-        // an expensive operation).
+//      We cannot easily check if status changed in multi-shop context, because product is loaded from a single shop
+//      (it would end up checking one shop product and leaving all other shops unhandled)
+//      So in multi-shop context we always reindex product
+        if ($shopConstraint->forAllShops() || $shopConstraint->getShopGroupId()) {
```

Comment:
```suggestion
        if ($shopConstraint->forAllShops()) {
```

Reply:
same.
only option if you want to get rid of this check is adding additional assertion in command construct, so we know that group constraint is impossible here.